core/injector/application_injector.cpp

```diff
@@ -4,3 +4,1305 @@
+      common::raise(batch.error());
+    }
+
+    auto log = log::createLogger("injector", "kagome");
+
+    for (const auto &[key, val] : genesis_raw_configs) {
+      log->debug("Key: {}, Val: {}", key.toHex(), val.toHex().substr(0, 200));
+      if (auto res = batch.value()->put(key, val); not res) {
+        common::raise(res.error());
+      }
+    }
+    if (auto res = batch.value()->commit(); not res) {
+      common::raise(res.error());
+    }
+
+    initialized = trie_storage;
+    return trie_storage;
+  }
+
+  sptr<storage::BufferStorage> get_level_db(
+      application::AppConfiguration const &app_config,
+      sptr<application::ChainSpec> chain_spec) {
+    static auto initialized =
+        boost::optional<sptr<storage::BufferStorage>>(boost::none);
+    if (initialized) {
+      return initialized.value();
+    }
+    auto options = leveldb::Options{};
+    options.create_if_missing = true;
+    auto db = storage::LevelDB::create(
+        app_config.databasePath(chain_spec->id()), options);
+    if (!db) {
+      auto log = log::createLogger("injector", "kagome");
+      log->critical("Can't create LevelDB in {}: {}",
+                    fs::absolute(app_config.databasePath(chain_spec->id()),
+                                 fs::current_path())
+                        .native(),
+                    db.error().message());
+      exit(EXIT_FAILURE);
+    }
+    initialized = db.value();
+    return initialized.value();
+  }
+
+  std::shared_ptr<application::ChainSpec> get_genesis_config(
+      application::AppConfiguration const &config) {
+    static auto initialized =
+        boost::optional<sptr<application::ChainSpec>>(boost::none);
+    if (initialized) {
+      return initialized.value();
+    }
+    auto const &genesis_path = config.genesisPath();
+
+    auto genesis_config_res =
+        application::ChainSpecImpl::loadFrom(genesis_path.native());
+    if (genesis_config_res.has_error()) {
+      common::raise(genesis_config_res.error());
+    }
+    initialized = genesis_config_res.value();
+    return genesis_config_res.value();
+  }
+
+  sptr<primitives::BabeConfiguration> get_babe_configuration(
+      sptr<runtime::BabeApi> babe_api) {
+    static auto initialized =
+        boost::optional<sptr<primitives::BabeConfiguration>>(boost::none);
+    if (initialized) {
+      return *initialized;
+    }
+
+    auto configuration_res = babe_api->configuration();
+    if (not configuration_res) {
+      common::raise(configuration_res.error());
+    }
+    auto log = log::createLogger("injector", "kagome");
+    for (const auto &authority :
+         configuration_res.value().genesis_authorities) {
+      log->debug("Babe authority: {}", authority.id.id.toHex());
+    }
+    configuration_res.value().leadership_rate.first *= 3;
+    initialized = std::make_shared<primitives::BabeConfiguration>(
+        configuration_res.value());
+    return *initialized;
+  }
+
+  consensus::SlotsStrategy get_slots_strategy(
```

Comment:
Possible is better to inject `app_config` and initialize `consensus::SlotsStrategy` in-place

Reply:
This way it won't be possible to instantiate classes dependent on SlotsStrategy with injector. Or do you mean changing the classess themselves so they depend on app config?

Reply:
Yep! Second one.